### PR TITLE
Integration tests setup

### PR DESCRIPTION
**Pull request by crypto-facs**
## Description

Started migrating some parts from the setup @yihuang had done in cronos to better test RPC endpoints with multiple configurations.

Tests are written in python to take advantage of pystarport for node configuration and also because of the extensive w3 library support for easy testing. `Nix` was used to manage reproducible builds and easy composability of dependencies.

This configurations currently allows to setup a basic ethermint and geth node and expect tests to pass on both networks. At the moment I just added a super simple `test_basic` to prove the concept as the configuration itself is very complex. 

Openning as a first draft to get some comments on it and as a foundation on which to build a robust test suite to maintain better reliabily on the RPC nodes. I will add extensive tests once this PR is merged and also other several configurations on which to tests like `pruned_nodes`, which should be straightforward with this suite as it has been done in Cronos. 

Big shoutout to @yihuang for setting this up

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: integration_tests/network.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+class Geth:
+    def __init__(self, w3):
+        self.w3 = w3
+
+
+def setup_ethermint(path, base_port, enable_auto_deployment=True):
+    cfg = Path(__file__).parent / (
+        "../scripts/ethermint-devnet.yaml"
+        if enable_auto_deployment
```
Comment (yihuang) [inline integration_tests/network.py:66]: we can remove `enable_auto_deployment` in ethermint.

File: result [hunk 2]
```diff
@@ -0,0 +1 @@
+/nix/store/n3b87plph97v8cx3n8d0laclbippp1nj-python3-3.9.13-env
```
Comment (yihuang) [inline result:1]: Should be put into `.gitignore`.

**Review by yihuang — CHANGES_REQUESTED** [review]
LGTM, left small cleanup suggestions.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -102,6 +102,28 @@ jobs:
           make test-integration
         if: env.GIT_DIFF
 
+  integration_tests:
```
Comment (fedekunze) [inline .github/workflows/test.yml:105]: we should use the `get-diff-action` to only run the integration tests when the go files have been updated (see above)

File: scripts/ethermint-devnet.yaml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+          max_deposit_period: "10s"
+          min_deposit:
+            - denom: "basetphoton"
+              amount: "1"
+      transfer:
+        params:
+          receive_enabled: true
+          send_enabled: true
+      feemarket:
+        params:
```
Comment (fedekunze) [inline scripts/ethermint-devnet.yaml:53]: do we need to add the other params?

**Review by fedekunze — COMMENTED** [review]
@crypto-facs can you move the tests to the existing `tests/` directory?

File: scripts/ethermint-devnet.yaml (hunk 4 above)
Comment (yihuang) [reply scripts/ethermint-devnet.yaml:53]: It's get merged with default parameters, so only need to specify the differences.

File: scripts/ethermint-devnet.yaml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+  genesis:
+    consensus_params:
+      block:
+        max_bytes: "1048576"
+        max_gas: "81500000"
+    app_state:
+      evm:
+        params:
+          evm_denom: basetphoton
```
Comment (ramacarlucho) [inline scripts/ethermint-devnet.yaml:39]: use `aphoton` as evm_denom in this file

File: result (hunk 2 above)
Comment (yihuang) [inline result:1]: should remove this file?

**Review by yihuang — APPROVED** [review]
LGTM

File: Makefile [hunk 6]
```diff
@@ -320,6 +320,12 @@ test-rpc:
 test-integration:
```
Comment (fedekunze) [inline Makefile:320]: are we using these? If not, I guess we can remove them and rename the other to `test-integration`

File: Makefile [hunk 7]
```diff
@@ -320,6 +320,12 @@ test-rpc:
...
 	./scripts/integration-test-all.sh -t "integration" -q 1 -z 1 -s 2 -m "integration" -r "true"
 
+run-integration-tests:
```
Comment (fedekunze) [inline Makefile:323]: nit, should be called `test-...`

File: nix/go-ethereum.nix [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  # A list of binaries to put into separate outputs
+  bins = [
+    "geth"
+    "clef"
+  ];
+
+in
+buildGoModule rec {
+  pname = "go-ethereum";
+  version = "1.10.15";
```
Comment (fedekunze) [inline nix/go-ethereum.nix:13]: shouldn't this be `1.10.19`?
```suggestion
  version = "1.10.19";
```

**Review by fedekunze — APPROVED** [review]

**Pull request merged**